source/Irrlicht/CGLTFMeshFileLoader.cpp

```diff
@@ -113,20 +115,134 @@ IAnimatedMesh* CGLTFMeshFileLoader::createMesh(io::IReadFile* file)
+			parent->GlobalMatrix.transformVect(vertex.Normal);
+			// Undo the translation, leaving us with scaling and rotation.
+			vertex.Normal -= parent->GlobalMatrix.getTranslation();
+			// Renormalize (length might have been affected by scaling).
+			vertex.Normal.normalize();
+		}
+		
+		auto maybeIndices = getIndices(meshIdx, j);
+		std::vector<u16> indices;
+		if (maybeIndices.has_value()) {
+			indices = std::move(maybeIndices.value());
+			for (u16 index : indices) {
+				if (index >= vertices->size())
+					throw std::runtime_error("index out of bounds");
+			}
+		} else {
+			// Non-indexed geometry
+			indices = std::vector<u16>(vertices->size());
+			for (std::size_t i = 0; i < vertices->size(); i += 3) {
+				// Reverse winding order per triangle
+				indices[i] = i + 2;
+				indices[i + 1] = i + 1;
+				indices[i + 2] = i;
+			}
+		}
+
+		auto *meshbuf = mesh->addMeshBuffer();
+		meshbuf->append(vertices->data(), vertices->size(),
+			indices.data(), indices.size());
+	}
+}
+
+// Base transformation between left & right handed coordinate systems.
+// This just inverts the Z axis.
+static core::matrix4 leftToRight = core::matrix4(
+	1, 0, 0, 0,
+	0, 1, 0, 0,
+	0, 0, -1, 0,
+	0, 0, 0, 1
+);
+static core::matrix4 rightToLeft = leftToRight;
+
+static core::matrix4 loadTransform(std::optional<std::variant<tiniergltf::Node::Matrix, tiniergltf::Node::TRS>> transform) {
+	if (!transform.has_value()) {
+		return core::matrix4();
+	}
+	core::matrix4 mat;
+	if (std::holds_alternative<tiniergltf::Node::Matrix>(*transform)) {
+		const auto &m = std::get<tiniergltf::Node::Matrix>(*transform);
+		// Note: Under the hood, this casts these doubles to floats.
+		mat = core::matrix4(
+			m[0], m[1], m[2], m[3],
+			m[4], m[5], m[6], m[7],
+			m[8], m[9], m[10], m[11],
+			m[12], m[13], m[14], m[15]);
+		// glTF uses column-major order, Irrlicht uses row-major order.
+		mat = mat.getTransposed();
+	} else {
+		const auto &trs = std::get<tiniergltf::Node::TRS>(*transform);
+		const auto &trans = trs.translation;
+		const auto &rot = trs.rotation;
+		const auto &scale = trs.scale;
+		core::matrix4 transMat;
+		transMat.setTranslation(core::vector3df(trans[0], trans[1], trans[2]));
+		core::matrix4 rotMat = core::quaternion(rot[0], rot[1], rot[2], rot[3]).getMatrix();
+		core::matrix4 scaleMat;
+		scaleMat.setScale(core::vector3df(scale[0], scale[1], scale[2]));
+		mat = transMat * rotMat * scaleMat;
+	}
+	return rightToLeft * mat * leftToRight;
+}
+
+void CGLTFMeshFileLoader::MeshExtractor::loadNode(
+		const std::size_t nodeIdx,
+		CSkinnedMesh* mesh,
+		CSkinnedMesh::SJoint *parent) const
+{
+	const auto &node = m_model.nodes->at(nodeIdx);
+	auto *joint = mesh->addJoint(parent);
+	const core::matrix4 transform = loadTransform(node.transform);
+	joint->LocalMatrix = transform;
+	joint->GlobalMatrix = parent ? parent->GlobalMatrix * joint->LocalMatrix : joint->LocalMatrix;
+	if (node.name.has_value()) {
+		joint->Name = node.name->c_str();
+	}

```

Comment:
It might be bad if there is no name. Maybe we need to throw an exception.

Reply:
Not sure about this. I don't really want to throw an exception; that would unnecessarily restrict the set of models we support. I'll look into Irrlicht to see if this currently poses a problem; maybe I'll refactor the code to use a `std::optional<std::string>`.

We could also consider hacks such as generating bone names (but those are, well, hacks).

Reply:
https://github.com/minetest/irrlicht/pull/284 was merged now.